File: gcd.pas
{
FreePascal: https://www.freepascal.org/
Tested with FreePascal 3.2

Usage:

  $ fpc gcd.pas -ogcd-pas
  $ ./gcd-pas 44 55 66 121
  11

}

Program GCD(output);

Uses sysutils; {StrToInt}

Function gcd2(a: integer; b: integer): integer;
Var c: integer;
Begin
  While b <> 0 Do
    Begin
      c := b;
      b := a Mod b;
      a := c;
    End;
  gcd2 := a;
End;

Function gcdn(n: Array Of integer): integer;
Var i: integer;
Begin
  gcdn := n[0];
  For i := 1 To High(n) Do {See also <http://wiki.freepascal.org/for-in_loop> }
    gcdn := gcd2(gcdn, n[i]);
End;


Var
  n: array Of integer;
  i: integer;
Begin
  If ParamCount = 0 then exit;

  SetLength(n, ParamCount);

  For i := 1 To ParamCount Do
    n[i-1] := StrToInt(ParamStr(i));

  Writeln(gcdn(n))
End.

File: gcd.hs
{-

Usage:

  $ runhaskell gcd.hs 22 11
  11

or:

  $ ghc gcd.hs -o gcd-hs
  $ ./gcd-hs 11 22 121
  11

-}
import System.Environment (getArgs)

gcd2 :: Integral a => a -> a -> a
gcd2 a 0 = a
gcd2 a b = gcd2 b (a `rem` b)

gcdn :: Integral a => [a] -> a
gcdn = foldl1 gcd2

out :: [String] -> IO ()
out [] = return ()
out a = print (gcdn (map read a) :: Integer)

main :: IO ()
main = getArgs >>= out
